staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -92,34 +93,49 @@ func validator(s *schema.Structural, isResourceRoot bool) *Validator {
 }
 
 // Validate validates all x-kubernetes-validations rules in Validator against obj and returns any errors.
-func (s *Validator) Validate(fldPath *field.Path, sts *schema.Structural, obj interface{}) field.ErrorList {
+func (s *Validator) Validate(fldPath *field.Path, sts *schema.Structural, obj interface{}, costBudget int64) (errs field.ErrorList, remainingBudget int64) {
+	remainingBudget = costBudget
 	if s == nil || obj == nil {
-		return nil
+		return nil, remainingBudget
 	}
 
-	errs := s.validateExpressions(fldPath, sts, obj)
+	errs, remainingBudget = s.validateExpressions(fldPath, sts, obj, remainingBudget)
+	if remainingBudget < 0 {
+		return errs, remainingBudget
+	}
 	switch obj := obj.(type) {
 	case []interface{}:
-		return append(errs, s.validateArray(fldPath, sts, obj)...)
+		var arrayErrs field.ErrorList
+		arrayErrs, remainingBudget = s.validateArray(fldPath, sts, obj, remainingBudget)
+		errs = append(errs, arrayErrs...)
+		return errs, remainingBudget
 	case map[string]interface{}:
-		return append(errs, s.validateMap(fldPath, sts, obj)...)
+		var mapErrs field.ErrorList
+		mapErrs, remainingBudget = s.validateMap(fldPath, sts, obj, remainingBudget)
+		errs = append(errs, mapErrs...)
+		return errs, remainingBudget
 	}
-	return errs
+	return errs, remainingBudget
 }
 
-func (s *Validator) validateExpressions(fldPath *field.Path, sts *schema.Structural, obj interface{}) (errs field.ErrorList) {
+func (s *Validator) validateExpressions(fldPath *field.Path, sts *schema.Structural, obj interface{}, costBudget int64) (errs field.ErrorList, remainingBudget int64) {
+	remainingBudget = costBudget
 	if obj == nil {
 		// We only validate non-null values. Rules that need to check for the state of a nullable value or the presence of an optional
 		// field must do so from the surrounding schema. E.g. if an array has nullable string items, a rule on the array
 		// schema can check if items are null, but a rule on the nullable string schema only validates the non-null strings.
-		return nil
+		return nil, remainingBudget
 	}
 	if s.compilationErr != nil {
 		errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("rule compiler initialization error: %v", s.compilationErr)))
-		return errs
+		return errs, remainingBudget
 	}
 	if len(s.compiledRules) == 0 {
-		return nil // nothing to do
+		return nil, remainingBudget // nothing to do
+	}
+	if remainingBudget == 0 {
```

Comment:
should this also error if it's already negative?
```suggestion
	if remainingBudget <= 0 {
```